std/macho.zig

```diff
@@ -0,0 +1,177 @@
+            const curr = &self.symbols[mid];
+            const next = &self.symbols[mid + 1];
+            if (address >= next.address) {
+                min = mid + 1;
+            } else if (address < curr.address) {
+                max = mid;
+            } else {
+                return curr;
+            }
+        }
+        return null;
+    }
+};
+
+pub fn loadSymbols(allocator: &mem.Allocator, in: &io.FileInStream) !SymbolTable {
+    var file = in.file;
+    try file.seekTo(0);
+
+    var hdr: MachHeader64 = undefined;
+    try readNoEof(in, &hdr);
+    if (hdr.magic != MH_MAGIC_64) return error.MissingDebugInfo;
+    const is_pie = MH_PIE == (hdr.flags & MH_PIE);
+
+    var pos: usize = @sizeOf(@typeOf(hdr));
+    var ncmd: u32 = hdr.ncmds;
+    while (ncmd != 0) : (ncmd -= 1) {
+        try file.seekTo(pos);
+        var lc: LoadCommand = undefined;
+        try readNoEof(in, &lc);
+        if (lc.cmd == LC_SYMTAB) break;
+        pos += lc.cmdsize;
+    } else {
+        return error.MissingDebugInfo;
+    }
+
+    var cmd: SymtabCommand = undefined;
+    try readNoEof(in, &cmd);
+
+    try file.seekTo(cmd.symoff);
+    var syms = try allocator.alloc(Nlist64, cmd.nsyms);
+    defer allocator.free(syms);
+    try readNoEof(in, syms);
+
+    try file.seekTo(cmd.stroff);
+    var strings = try allocator.alloc(u8, cmd.strsize);
+    errdefer allocator.free(strings);
+    try in.stream.readNoEof(strings);
+
+    var nsyms: usize = 0;
+    for (syms) |sym| if (isSymbol(sym)) nsyms += 1;
+    if (nsyms == 0) return error.MissingDebugInfo;
+
+    var symbols = try allocator.alloc(Symbol, nsyms + 1);  // Room for sentinel.
+    errdefer allocator.free(symbols);
+
+    var pie_slide: usize = 0;
+    var nsym: usize = 0;
+    for (syms) |sym| {
+        if (!isSymbol(sym)) continue;
+        const start = sym.n_strx;
+        const end = ??mem.indexOfScalarPos(u8, strings, start, 0);
+        const name = strings[start..end];
+        const address = sym.n_value;
+        symbols[nsym] = Symbol { .name = name, .address = address };
+        nsym += 1;
+        if (is_pie and mem.eql(u8, name, "_SymbolTable_deinit")) {
+            pie_slide = @ptrToInt(SymbolTable.deinit) - address;
+        }
+    }
+
+    // Effectively a no-op, lld emits symbols in ascending order.
+    std.sort.insertionSort(Symbol, symbols[0..nsyms], Symbol.addressLessThan);
+
+    // Insert the sentinel.  Since we don't know where the last function ends,
+    // we arbitrarily limit it to the start address + 4 KB.
+    const top = symbols[nsyms - 1].address + 4096;
+    symbols[nsyms] = Symbol { .name = "", .address = top };
+
+    if (pie_slide != 0) {
+        for (symbols) |*symbol| symbol.address += pie_slide;
+    }
+
+    return SymbolTable {
+        .allocator = allocator,
+        .symbols = symbols,
+        .strings = strings,
+    };
+}
+
+fn readNoEof(in: &io.FileInStream, sink: var) !void {
+    if (@typeOf(sink) == []Nlist64) {
```

Comment:
I couldn't figure out a way to do this generically.

Reply:
I have a lot of responses to this, starting with minimal changes, ending with let's change the standard library.

1. I'm actually fine with this, given that it's at least correct for this file, and it's a private function so it's not really exposed to much of anything.

2. Here's how to do what you wanted to do:

```zig
fn readNoEof(in: &io.FileInStream, sink: var) !void {
    switch (@typeId(@typeOf(sink)) {
        builtin.TypeId.Slice => {
            const T = @typeOf(sink).Child;
            ...
        },
        builtin.TypeId.Pointer => {
            const T = @typeOf(sink).Child;
            ...
        },
        else => @compileError("expected slice or pointer"),
    }
}
```

3. However, let's use this as an example to write idiomatic zig code. I think it would be better to separate this out into 2 functions:
    
```zig
fn readNoEof(in: &io.FileInStream, comptime T: type, result: []T) !void {
    return in.stream.readNoEof(([]u8)(result));
}
fn readOneNoEof(in: &io.FileInStream, comptime T: type, result: &T) !void {
    return readNoEof(in, T, result[0..1]);
}
```
related: #287 

4. One more thing, this looks like it's a pretty common thing to want to do. Let's modify `std.io.InStream` to support these APIs. I think that it's ok to break backwards compatibility of readNoEof and give it a comptime parameter of the type you want to read. Existing callsites can be updated to get the old behavior by passing in `u8`.
    
    And then we can add a safety check, if you pass a struct to #183, it will give you a compile error if your struct is not `packed` (because you would get an undefined value when deserializing a non-packed struct).
    
    I think this macho code should be passing around a `&std.io.InStream` instead of the `&std.io.FileInStream` directly. This is where that `readNoEof` API would go. I understand that it's a little weird because you'd have to use `var` for the parameter due to error sets. That's #764 and we don't have a good solution for that yet (other than "just use var").

5. Finally, I think we should go back to (3) for now, because in order for (4) to make sense, we need the ability to specify endianness in packed structs. That's pending #649. Once packed structs support the concept of endianness, it makes sense to use them as types to serialization/deserialization functions.

Hope that helps!

My proposal is:
 * do (3) since it's easy
 * merge this PR since it's better than status quo
 * I'll bump the priority of #649 
 * Once we get packed struct endianness working, we can add support for serializing/deserializing them using `std.io.OutStream` / `std.io.InStream` and then update this code to use `std.io.InStream` instead of `std.io.FileInStream`